src/biotite/sequence/profile.py

```diff
@@ -280,3 +291,88 @@ def _general_to_consensus(self):
+        and the given pseudocount.
+        This new matrix has the same shape as 'symbols'.
+
+        Parameters
+        ----------
+        pseudocount: int
+            Amount added to the number of observed cases in order to change
+            the expected probability.
+            (Default: 0)
+
+        Returns
+        -------
+        ppm: ndarray, dtype=float, shape=(n,k)
+            Calculated the position probability matrix.
+        """
+        if pseudocount < 0:
+            raise ValueError(
+                f"Pseudocount can not be smaller than zero."
+            )
+        return (self.symbols + pseudocount / len(self.symbols)) / \
+               (np.sum(self.symbols, axis=1)[:, np.newaxis] + pseudocount)
+
+    def log_odds_matrix(self, background_frequencies=None, pseudocount=0):
+        """
+        Calculate the position weight matrix (pwm) based on the
+        position probability matrix (ppm) (with given pseudocount) and
+        background_frequencies.
+        This new matrix has the same shape as 'symbols'.
+
+        Parameters
+        ----------
+        pseudocount: int
+            Amount added to the number of observed cases in order to change
+            the expected probability.
+            (Default: 0)
+        background_frequencies: float
+            (Default: 1/len(alphabet))
+        Returns
+        -------
+        pwm: ndarray, dtype=float, shape=(n,k)
+            Calculated the position weight matrix.
+        """
+        if background_frequencies is None:
+            background_frequencies = 1 / len(self.alphabet)
+        ppm = self.probability_matrix(pseudocount)
+        return np.log2(ppm / background_frequencies)
+
+    def sequence_probability_from_matrix(self, sequence, matrix):
+        """
+        Calculate probability of a sequence based on either the
+        position probability matrix (ppm) or the position weight matrix (pwm).
+
+        Parameters
+        ----------
+        sequence : Sequence
+           The input sequence.
+        matrix : ndarray, dtype=float, shape=(n,k)
+            The chosen position matrix to use to calculate the probability
+            of the input sequence (either ppm or pwm).
+
+        Returns
+        -------
+        probability: float
+           The calculated probability for the input sequence based on
+           the chosen position matrix.
+        """
+        if len(sequence) != len(matrix):
+            raise ValueError(
+                f"The given sequence has a different length ({len(sequence)}) than "
+                f"the matrix ({len(matrix)})."
+            )
+        if not matrix.shape == self.symbols.shape:
+            raise ValueError(
+                f"Matrix {matrix.shape} must be of same shape "
+                f"as 'symbols' {self.symbols.shape}"
+            )
+        probability = 1
+        codes = sequence.code
+        for i in range(len(codes)):
+            probability = probability * matrix[i, codes[i]]
```

Comment:
I think this loop can be vectorized with `numpy.prod(matrix[np.arange(len(sequence)), sequence.code])`